File: src/main/java/util/Notepad/WinNotepad.java
package util.Notepad;

import java.awt.*;
import java.awt.event.KeyEvent;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import lombok.experimental.ExtensionMethod;
import lombok.extern.log4j.Log4j2;
import util.Extension.RobotExtension;

@Log4j2
@ExtensionMethod({RobotExtension.class})
public class WinNotepad implements Notepad {
	public static final String home = System.getProperty("user.home") + File.separator + "ParseButPro";
	private Robot r;
	private Process process;

	WinNotepad() {
		try {
			r = new Robot();
		} catch (AWTException e) {
			log.error("Failed to initialize Robot", e);
		}
	}

	public void openNotepad() throws IOException {
		Files.createDirectories(Paths.get(home));
		ProcessBuilder processBuilder = new ProcessBuilder("C:\\Program Files\\Notepad++\\notepad++.exe");
		process = processBuilder.start();
		r.delay(1000);
	}

	public void writeText(String text) {
		r.type(text);
	}

	public void deleteText() {
		r.control(KeyEvent.VK_A);
		r.type(KeyEvent.VK_BACK_SPACE);
	}

	public void addNewLine() {
		r.enter();
	}

	public void saveFileAs(String name) {
		r.control(KeyEvent.VK_S);
		r.delay(1000);
		r.type(name);
		for (int i = 0; i < 7; i++) {
			r.type(KeyEvent.VK_TAB);
		}
		r.enter();
		r.type(home);
		r.enter();
		r.delay(500);
		r.enter();
		r.delay(500);
		r.enter();
		r.delay(500);
		r.enter();
	}

	public void openNewFile() {
		r.control(KeyEvent.VK_N);
	}

	public void closeFile() {
		r.control(KeyEvent.VK_W);
	}

	public void closeNotepad() throws InterruptedException {
		r.control(KeyEvent.VK_F4);
		log.info("Exited Notepad++ with code: {}", process.waitFor());
	}
}

File: src/main/kotlin/util/notepad/WinNotepad.kt
package util.notepad

import util.audio.NativeTTS
import util.extension.control
import util.extension.enter
import util.extension.type
import java.awt.Robot
import java.awt.event.KeyEvent
import java.io.File.separator
import java.nio.file.Files
import java.nio.file.Paths

class WinNotepad : Notepad {
    private val homeDirectory = System.getProperty("user.home") + separator + "Aries"
    private val robot = Robot()
    private var process: Process? = null

    init {
        Files.createDirectories(Paths.get(homeDirectory))
    }

    override fun openNotepad() {
        process = ProcessBuilder("C:\\Program Files\\Notepad++\\notepad++.exe").start()
        robot.delay(1000)
    }

    override fun deleteText() {
        robot.apply {
            control(KeyEvent.VK_A)
            type(KeyEvent.VK_BACK_SPACE)
        }
    }

    override fun addNewLine() = robot.enter()

    override fun saveFileAs(name: String) {
        robot.apply {
            println("saving file as $name")
            control(KeyEvent.VK_S)
            delay(1000)
            type(name)
            repeat(7) { type(KeyEvent.VK_TAB) }
            enter()
            type("${System.getProperty("user.home")}${separator}Documents$separator")
            repeat(10) { type(KeyEvent.VK_TAB) }
            enter()
        }
    }

    override fun openNewFile() = robot.control(KeyEvent.VK_N)

    override fun closeNotepad() {
        process?.let {
            it.destroy()
            val exitCode = it.waitFor()
            val message =
                if (exitCode == 1) {
                    "Exited Notepad++"
                } else {
                    "Please open Notepad++ with the open notepad command"
                }
            NativeTTS.tts(message)
        } ?: run { NativeTTS.tts("Notepad++ is not open") }
        process = null
    }
}
